drafts/rxjs-memory-leaks.md

```diff
@@ -0,0 +1,223 @@
+      this.searchService.updateSearchResults(searchTerm);
+    });
+  }
+}
+```
+
+```xml
+<mat-form-field appearance="fill">
+  <mat-label>Search</mat-label>
+  <input matInput [formControl]="searchCtrl" (focus)="onSearchFieldFocus()" />
+</mat-form-field>
+```
+
+The `valueChanges` returns an Observable of type _any_. It emits an event when the value of the search control changes, so, with every new user keystroke. We use the [RxJs operator](https://rxjs.dev/api/operators/debounceTime) `debounceTime` to only emit a value when there is no user input for at least 1s. In our callback, we simply call the backend service to update the search results.
+
+**Identifying the Memory Leak**  
+Now, when we subscribe to the search input changes, a subscription is created in the heap memory. Currently, we do not unsubscribe from the `valueChanges` Observable.  
+As soon as the `SearchComponent` is destroyed, the subscription is still active and not cleaned up by the garbage collector. This becomes a problem when the `SearchComponent` is created and destroyed frequently, e.g. when the search input is only displayed after clicking a search button before. In that case, more and more heap space is occupied by those obsolete `valueChanges` subscriptions. The browser will become slower and slower and the search feature unusable. We have a memory leak.
+
+A page refresh would clear the heap memory and reset the problem state. Making it difficult to identify issues like that during development with frequent re-compilations and page refreshes.
+
+Luckily there are various techniques to deal with memory leaks. In the following, we present the most popular ones. Additionally, we outline the extension of the linting rules to prevent memory leaks before they occur.
+
+## Fixing Memory Leaks
+
+There are various techniques to fix memory leaks. We abstain from presenting third-party packages that deal with this issue as oftentimes this is not an option in corporate projects.
+
+Here are the most common ones.
+
+1. **ngOnDestroy**
+2. **Scalable ngOnDestroy**
+3. **Mixin**
+
+**1\. ngOnDestroy**  
+A straightforward approach is to simply unsubscribe from all Observables within a component before it gets destroyed. It is recommended to do this in the `ngOnDestroy` lifecycle hook. To that end, we assign the subscription to a variable `searchTerm$` and unsubscribe from this subscription when the component gets destroyed.
+
+```typescript
+@Component({
+  selector: 'app-search',
+  templateUrl: './search.component.html',
+  styleUrls: ['./search.component.scss'],
+})
+export class SearchComponent implements OnInit, OnDestroy {
+  searchCtrl: FormControl = new FormControl();
+  searchTerm$: Subscription;
+
+  ngOnInit(): void {
+    this.searchTerm$ = this.searchCtrl.valueChanges
+      .pipe(debounceTime(1000))
+      .subscribe((searchTerm: string) => {
+        ...
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.searchTerm$.unsubscribe();
```

Comment:
I'd typcially use `this.searchTerm$?.unsubscribe();`
I've seen weird short render cycles on comonents so I'm a bit paranoid about potential null pointers because the "ngOnInit" did not run

Reply:
same 👍  but I didn't want to open up a whole new topic here ;)